### PR TITLE
support static NVFP4 HF export

### PR DESCRIPTION
**Pull request by Fridah-nv**
## What does this PR do?

**Type of change:** ? new feature

**Overview:** ?
Supports export `NVFP4StaticQantizer` in unified huggingface checkpoint, as a deployment path for PTQ algorithms such as MSE

## Usage

```python
# checkpoint generation
python examples/llm_ptq/hf_ptq.py --pyt_ckpt_path Qwen/Qwen3-8B  --qformat nvfp4_mse --export_path test-Qwen3-8B-Instruct-MSE-FP8-sweep-FP4 --kv_cache_qformat none --trust_remote_code 

```

## Testing
Tested generated Qwen3  8B checkpoint with trtllm serve and nv_eval example in `Model-Optimizer-Internal/examples/nv_eval`.

NV eval results:
```
| Groups |Version|Filter|n-shot|  Metric   |   |Value |   |Stderr|
|--------|-------|------|------|-----------|---|-----:|---|-----:|
|mmlu_str|       |none  |      |exact_match|↑  |0.7186|±  |0.0036|
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

* **New Features**
  * Added support for static NVFP4 quantizers that utilize pre-computed calibration scales.
  * Introduced new NVFP4 W4A4 quantization configuration with optional FP8 scale sweep.

* **Performance Improvements**
  * Static quantizers now skip unnecessary dynamic scaling factor recalculation.
  * Unified quantization handling for improved consistency and efficiency.

### CODE REVIEW

File: modelopt/torch/quantization/config.py [hunk 1]
```diff
@@ -388,6 +388,32 @@
     "algorithm": "max",
 }
 
+NVFP4_WEIGHT_MSE_FP8_SWEEP_CFG = {
```
Comment (realAsma) [inline modelopt/torch/quantization/config.py]: Activation is quantized, can we make it more evident?
```suggestion
NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG = {
```

File: modelopt/torch/export/quant_utils.py [hunk 2]
```diff
@@ -299,6 +299,31 @@ def get_weight_scaling_factor(module: nn.Module, weight_name: str = "weight") ->
...
+        weight_scaling_factor_2 = global_amax / (6.0 * 448.0)
+        per_block_scale = per_block_amax / (6.0 * weight_scaling_factor_2.to(per_block_amax.device))
+        per_block_scale[per_block_scale == 0] = 1.0
+
+        # Reshape per_block_scale to match weight's block structure: (rows, num_blocks_per_row)
+        num_blocks_per_row = weight.shape[-1] // block_size
+        expected_shape = (*weight.shape[:-1], num_blocks_per_row)
+        per_block_scale = per_block_scale.view(expected_shape)
+
+        return per_block_scale.to(torch.float8_e4m3fn)
```
Comment (realAsma) [inline modelopt/torch/export/quant_utils.py]: nit:
This is the way I think about FP8 quantization of scale. I think it is more intuitive as following

```suggestion
        per_block_scale_max = global_amax.float() / 6.0 # importance do the scale in float
        per_block_scale = per_block_amax.float() / 6.0
        per_block_scale[per_block_scale == 0] = 1.0

        # Reshape per_block_scale to match weight's block structure: (rows, num_blocks_per_row)
        num_blocks_per_row = weight.shape[-1] // block_size
        expected_shape = (*weight.shape[:-1], num_blocks_per_row)
        per_block_scale = per_block_scale.view(expected_shape)

        per_block_scale_fp8 = (per_block_scale * 448.0 / per_block_scale_max).to(torch.float8_e4m3fn)
        return per_block_scale_fp8
```

File: modelopt/torch/export/quant_utils.py [hunk 3]
```diff
@@ -299,6 +299,31 @@ def get_weight_scaling_factor(module: nn.Module, weight_name: str = "weight") ->
...
+    # Handle NVFP4 static quantizer (has pre-computed per-block amax values)
+    is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer) or getattr(
+        weight_quantizer, "_is_nvfp4_static_quantizer", False
+    )
+    if is_nvfp4_static:
+        # For static NVFP4, _amax contains per-block amax values and _global_amax is the global scale
+        assert (
+            hasattr(weight_quantizer, "_global_amax") and weight_quantizer._global_amax is not None
+        )
+        global_amax = weight_quantizer._global_amax.float()
```
Comment (realAsma) [inline modelopt/torch/export/quant_utils.py]: Can we use this
```suggestion
        global_amax = weight_quantizer.global_amax.float()
```

File: modelopt/torch/export/quant_utils.py [hunk 4]
```diff
@@ -299,6 +299,31 @@ def get_weight_scaling_factor(module: nn.Module, weight_name: str = "weight") ->
...
+
+    # Handle NVFP4 static quantizer (has pre-computed per-block amax values)
+    is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer) or getattr(
+        weight_quantizer, "_is_nvfp4_static_quantizer", False
+    )
+    if is_nvfp4_static:
+        # For static NVFP4, _amax contains per-block amax values and _global_amax is the global scale
+        assert (
+            hasattr(weight_quantizer, "_global_amax") and weight_quantizer._global_amax is not None
+        )
```
Comment (realAsma) [inline modelopt/torch/export/quant_utils.py]:
```suggestion
        assert weight_quantizer.global_amax is not None
```

File: modelopt/torch/export/quant_utils.py [hunk 5]
```diff
@@ -1315,6 +1350,22 @@ def preprocess_linear_fusion(modules: list[torch.nn.Module], resmooth_only=False
...
+                for m in modules
+                if hasattr(m.weight_quantizer, "_global_amax")
+                and m.weight_quantizer._global_amax is not None
+            ]
+            if global_amax_list:
+                unified_global_amax = torch.max(torch.stack(global_amax_list))
+                for module in modules:
+                    if hasattr(module.weight_quantizer, "_global_amax"):
+                        module.weight_quantizer._global_amax = unified_global_amax
+
```
Comment (Fridah-nv) [inline modelopt/torch/export/quant_utils.py:1405]: TRTLLM requires fusible linear layers to have same `weight_scale_2` , so here I set global_amax to be the max value across all modules being fused. It's also what the dynamic path is doing.
After discussing with cursor, I think this won't have much impact to accuracy:
```
The quantization math:
weights_scaling_factor_2 = _global_amax / (6.0 * 448.0)
weights_scaling_factor = _amax / (6.0 * wsf2) = _amax * 448.0 / _global_amax
Effective scale used in quantization: weights_scaling_factor * weights_scaling_factor_2 = _amax / 6.0
``` 

> Since the effective scale is wsf * wsf2 = _amax / 6.0, the _amax values are determined purely by the weight values themselves through MSE calibration, not by _global_amax.

>  The _global_amax is just a separate parameter that determines how to split the representation between the two scale factors (wsf and wsf2), but since they multiply together and _global_amax cancels out, it has no effect on the actual quantization.

I think it would still have some effect since wsf and wsf2 are kept in different precision, but the gap should not be large. @realAsma Please let me know if it makes sense to you.

File: modelopt/torch/export/unified_export_hf.py [hunk 6]
```diff
@@ -486,11 +490,20 @@ def _export_quantized_weight(
...
         )
-        weight_scale = NVFP4QTensor.get_weights_scaling_factor(
-            weight,
-            block_size=block_size,
-            weights_scaling_factor_2=weight_scale_2,
-        )[0]
+
+        # Check if this is a static NVFP4 quantizer (has pre-computed scales from MSE calibration)
+        # For static NVFP4, weight_scale is already computed from static _amax values in get_weight_scaling_factor
+        is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer) or getattr(
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py]: QQ: why isinstance(weight_quantizer, NVFP4StaticQuantizer) is not good enough?

File: modelopt/torch/export/quant_utils.py [hunk 7]
```diff
@@ -343,6 +368,16 @@ def get_weight_scaling_factor_2(module: nn.Module, weight_name: str = "weight")
 
     quantization_format = get_quantization_format(module)
 
+    # Handle NVFP4 static quantizer (use _global_amax instead of _amax)
```
Comment (cjluo-nv) [inline modelopt/torch/export/quant_utils.py]: what's the difference between _global_amax and _amax?

File: modelopt/torch/export/quant_utils.py [hunk 8]
```diff
@@ -299,6 +299,31 @@ def get_weight_scaling_factor(module: nn.Module, weight_name: str = "weight") ->
         return get_scaling_factor(weight_quantizer[0])
 
     quantization_format = get_quantization_format(module)
+
+    # Handle NVFP4 static quantizer (has pre-computed per-block amax values)
```
Comment (cjluo-nv) [inline modelopt/torch/export/quant_utils.py]: can we merge these logics to the nvfp4_tensor.py so we don't scatter nvfp4 related features in various places or have side branches.

Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py]: Thanks for the suggestion, I moved implementation details into NVFP4QTensor, cannot avoid all branches in quant_utils.py because static NVFP4 does not need re-calibration

File: modelopt/torch/export/quant_utils.py (hunk 7 above)
Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py]: For the static quantizer, _global_amax is the per-tensor global scale, _amax are per-block scales

File: modelopt/torch/export/unified_export_hf.py (hunk 6 above)
Comment (Fridah-nv) [reply modelopt/torch/export/unified_export_hf.py]: It should be enough, thanks for catching it!

**Comment by Fridah-nv** [comment]
> @Fridah-nv Could we add a sample command in the PR description and the MMLU results you have?

Added, thanks for reminding.

> Also, is the checkpoint format the same comparing static NVFP4 vs our default NVFP4?

Yes, I checked that `hf_quant_config.json` and `config.json` are the same for the two formats.

File: modelopt/torch/export/quant_utils.py [hunk 9]
```diff
@@ -299,16 +299,19 @@ def get_weight_scaling_factor(module: nn.Module, weight_name: str = "weight") ->
         return get_scaling_factor(weight_quantizer[0])
 
     quantization_format = get_quantization_format(module)
-    # If NVFP4, we need to return quantized per_block scaling factors
-    if quantization_format in [
+
+    # Handle NVFP4 variants (static or dynamic)
+    is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer)
+    if is_nvfp4_static or quantization_format in [
```
Comment (cjluo-nv) [inline modelopt/torch/export/quant_utils.py]: what's the quantization_format for NVFP4StaticQuantizer? do we need is_nvfp4_static here?

File: modelopt/torch/export/quant_utils.py [hunk 10]
```diff
@@ -343,27 +347,26 @@ def get_weight_scaling_factor_2(module: nn.Module, weight_name: str = "weight")
 
     quantization_format = get_quantization_format(module)
 
-    # Calibrate weight quantizer if amax is not set for all NVFP4 variants
-    if quantization_format in [
+    is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer)
+    if is_nvfp4_static or quantization_format in [
```
Comment (cjluo-nv) [inline modelopt/torch/export/quant_utils.py]: same

File: modelopt/torch/export/unified_export_hf.py [hunk 11]
```diff
@@ -502,11 +506,18 @@ def _export_quantized_weight(
...
-            weight,
-            block_size=block_size,
-            weights_scaling_factor_2=weight_scale_2,
-        )[0]
+
+        # Check if this is a static NVFP4 quantizer (has pre-computed scales from MSE calibration)
+        # For static NVFP4, weight_scale is already computed from static _amax values in get_weight_scaling_factor
+        is_nvfp4_static = isinstance(weight_quantizer, NVFP4StaticQuantizer)
+
+        if not is_nvfp4_static:
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py]: do we need to handle the else condition?

File: modelopt/torch/quantization/qtensor/nvfp4_tensor.py [hunk 12]
```diff
@@ -52,12 +52,87 @@ def get_e2m1_bounds(cls, device):
...
+            Tuple of (per_block_scale, weights_scaling_factor_2).
+        """
+        block_size = weight_quantizer.block_sizes[-1]
+
+        if weights_scaling_factor_2 is None:
+            weights_scaling_factor_2 = cls.get_weights_scaling_factor_2_from_quantizer(
+                weight_quantizer
+            )
+
+        if cls._is_static_quantizer(weight_quantizer):
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/qtensor/nvfp4_tensor.py:110]: so in static NVFP4 quant case:

You use _amax for per-block amax and global_amax for the per-tensor amax?

File: modelopt/torch/quantization/qtensor/nvfp4_tensor.py [hunk 13]
```diff
@@ -52,12 +52,87 @@ def get_e2m1_bounds(cls, device):
             cls.e2m1_bounds_on_device[device] = e2m1_bounds.to(device)
         return cls.e2m1_bounds_on_device[device]
 
+    @classmethod
+    def _is_static_quantizer(cls, weight_quantizer) -> bool:
+        """Check if the weight quantizer is a static NVFP4 quantizer with pre-computed amax."""
+        return hasattr(weight_quantizer, "global_amax") and weight_quantizer.global_amax is not None
+
     @classmethod
     def get_weights_scaling_factor_2_from_quantizer(cls, weight_quantizer):
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/qtensor/nvfp4_tensor.py:61]: could you add a unittest?

File: modelopt/torch/export/unified_export_hf.py [hunk 14]
```diff
@@ -502,11 +506,18 @@ def _export_quantized_weight(
...
+            # For dynamic NVFP4, compute scales from weights
+            weight_scale = NVFP4QTensor.get_weights_scaling_factor(
+                weight,
+                block_size=block_size,
+                weights_scaling_factor_2=weight_scale_2,
+            )[0]
```
[the remark this answers is not in the document]
Comment (cjluo-nv) [reply modelopt/torch/export/unified_export_hf.py]: @Fridah-nv could you fix this?

File: modelopt/torch/export/quant_utils.py (hunk 9 above)
Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py]: I add this branch because I think we don't need `_ensure_weight_quantizer_calibrated` for NVFP4 static case. All NVFP4StaticQuantizer should be calibrated by the time of export

Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py]: I'm not sure about the case for `_ensure_weight_quantizer_calibrated` to be effective. I understand activation quantizers might not be calibrated. But why weight quantizers might not be reached before export? If this is needed, I can ensure `_ensure_weight_quantizer_calibrated` work with NVFP4StaticQuantizer

File: modelopt/torch/quantization/qtensor/nvfp4_tensor.py (hunk 12 above)
Comment (Fridah-nv) [reply modelopt/torch/quantization/qtensor/nvfp4_tensor.py:110]: Yes, this is true

File: modelopt/torch/export/unified_export_hf.py (hunk 11 above)
Comment (Fridah-nv) [reply modelopt/torch/export/unified_export_hf.py]: The static NVFP4 case is already handled by L459-L462
```
        else:
            sub_module.register_buffer(
                quantizer_attrs.weight_scale, get_weight_scaling_factor(sub_module, weight_name)
            )
```
Which calls `NVFP4QTensor.get_weights_scaling_factor_from_quantizer` that handles both static and dynamic NVFP4 case.

Comment (Fridah-nv) [reply modelopt/torch/export/unified_export_hf.py]: These lines L511-L514 for handling dynamic NVFP4 could be removed potentially, I just need to double check on the dynamic BMM case

File: modelopt/torch/export/unified_export_hf.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Fridah-nv) [reply modelopt/torch/export/unified_export_hf.py]: We are not supporting static NVFP4 BMM yet, it won't be trigger in calibration algorithm, let me add a guard here

File: modelopt/torch/export/quant_utils.py (hunk 9 above)
Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py]: I understand the case that the weight quantizers of some experts might not be calibrated now. Updated _ensure_weight_quantizer_calibrated to handle NVFP4StaticQuantizer

File: modelopt/torch/export/unified_export_hf.py (hunk 11 above)
Comment (Fridah-nv) [reply modelopt/torch/export/unified_export_hf.py]: Update: removed this if branch and add a warning before BMM quantizers handling.

File: modelopt/torch/export/quant_utils.py (hunk 5 above)
Comment (cjluo-nv) [reply modelopt/torch/export/quant_utils.py:1405]: @Fridah-nv are you saying that. e.g. for QKV,

You keep the wsf1 not changed, but just pick the wsf2 of QKV to be the max of the individual wsf2?

File: modelopt/torch/export/unified_export_hf.py [hunk 15]
```diff
@@ -496,6 +500,11 @@ def _export_quantized_weight(
         expert_type in type(sub_module).__name__
         for expert_type in ["Llama4TextExperts", "GptOssExperts"]
     )
+    if is_bmm_expert_weight and isinstance(weight_quantizer, NVFP4StaticQuantizer):
+        warnings.warn(
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py]: how about just throw exception here? Prefer failing fast instead of unexpected accuracy bug.

File: CHANGELOG.rst [hunk 16]
```diff
@@ -32,6 +32,7 @@ NVIDIA Model Optimizer Changelog (Linux)
 - Add LTX-2 and Wan2.2 (T2V) support in the diffusers quantization workflow.
 - Add PTQ support for GLM-4.7, including loading MTP layer weights from a separate ``mtp.safetensors`` file and export as-is.
 - Add support for image-text data calibration in PTQ for Nemotron VL models.
+- Add support for advanced weight scale search for NVFP4 quantization and its export path.
```
Comment (cjluo-nv) [inline CHANGELOG.rst:36]: this is not a 0.42 feature. Please move to 0.43

File: modelopt/torch/export/quant_utils.py (hunk 5 above)
Comment (Fridah-nv) [reply modelopt/torch/export/quant_utils.py:1405]: Only the global_amax(wsf1) needs to be unified, wsf2 won't be updated. The concern above is that MSE algo ties wsf1 to wsf2, I'm worried that updating wsf1 only might affect accuracy. But it looks fine in math and in our experiements results.
